log-levels.md

```diff
@@ -0,0 +1,203 @@
+
+Logging `critical` logs is allowed for libraries, however as the name implies - only in the most critical situations. Most often this implies that the library will *stop functioning* after such log has been issued. End users are thought to expect that a logged critical error is _very_ important, and they may have set up their systems to page people in the middle of the night to investigate the production system _right now_ when such log statements are detected. So please be careful about logging these kinds of errors.
+
+Some libraries and situations may not be entirely clear with regards to what log level is "best" for them. In such situations, it sometimes is worth it to allow the end-users of the library to be able to configure the levels of specific groups of messages. You can see this in action in the Soto library [here](https://github.com/soto-project/soto-core/pull/423/files#diff-4a8ca7e54da5b22287900dd8cf6b47ded38a94194c1f0b544119030c81a2f238R649) where an `Options` object allows end users to configure the level at which requests are logged (`options.requestLogLevel`) which is then used as `log.log(self.options.requestLogLevel)`.
+
+#### Examples
+
+`trace` level logging:
+
+- could include various additional information about a request, such as various diagnostics about created datastructures, the state of caches or similar, which are created in order to serve a request.
+- could include "begin operation" and "end operation" logs, 
+  - however please do consider using swift-distributed-tracing to instrument such "begin/end" events as tracing can be more efficient than logging which may need to create string representations of the durations and log levels etc.
+
+`debug` level logging:
+
+- may include a single log statement for opening a connection, or accepting a request etc,
+- it can include a _high level_ overview of what is going on in the library "started work, processing step X, finished work X, result code 200"
+
+### Log levels to avoid
+
+All these rules are only _general_ guidelines, and as such may have exceptions.Consider the following examples and rationale for why logging at high log levels by a library may not be desirable:
+
+It is generally _not acceptable_ for a service client (e.g. an http client) to log an `error` when a request has failed. End-users may be using the client to probe if an endpoint is even responsive or not, and a failure to respond may be _expected_ behavior. Logging errors would only confuse and pollute their logs.
+Instead, libraries should either `throw`, or return an `Error` value that users of the library will have enough knowladge about if they should log or ignore it.
+
+It is even less acceptable for a library to log any successful operations. This leads to flodding server side systems, especially 
```

Comment:
sentence cuts off?

Reply:
Thanks fixed: 

> It is even less acceptable for a library to log any successful operations. This leads to flodding server side systems, especially if e.g. one were to log every successfully handled request in a server side application this can easily flood and overwhelm logging systems when deployed to production where many end users are connected to the same server. Such issues are rarely found in development time, because of only a single peer requesting things from the service-under-test.